tests/test_composition_functions.py

```diff
@@ -0,0 +1,102 @@
+import numpy as np
+import torch
+from scripts import tw_transform, transweigh_weight, concat
+import unittest
+
+
+class CompositionFunctionsTest(unittest.TestCase):
+    """
+    This class tests the composition functions.
+    This test suite can be ran with:
+        python -m unittest -q tests.CompositionFunctionsTest
+    """
+
+    def setUp(self):
+        self.u = torch.from_numpy(np.array([[1, 1, 1]], dtype='float32'))
+        self.v = torch.from_numpy(np.array([[1, 0, 0]], dtype='float32'))
+
+        self.transformations_tensor = np.full(shape=(2, 6, 3), fill_value=0.0, dtype='float32')
+        two_eyes = np.concatenate((np.eye(3), np.eye(3)), axis=0)
+        self.transformations_tensor[0] = np.copy(two_eyes)
+        self.transformations_tensor[1] = np.copy(two_eyes)
+        self.transformations_tensor = self.set_tensor_values(tensor=self.transformations_tensor, matrix=[0, 1, 1, 1],
+                                                             column=[2, 0, 1, 3], row=[2, 0, 1, 0], values=[0] * 4)
+        self.transformations_tensor = torch.from_numpy(self.transformations_tensor)
+
+        self.transformations_bias = torch.from_numpy(np.full(shape=(2, 3), fill_value=0.0, dtype='float32'))
+
+        self.combining_tensor = np.full(shape=(2, 3, 3), fill_value=0.0, dtype='float32')
+        matrix_numbers = [0] * 9 + [1] * 9
+        column_numbers = [0] * 3 + [1] * 3 + [2] * 3 + [0] * 3 + [1] * 3 + [2] * 3
+        row_numbers = [0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2, 0, 1, 2]
+        values = [1, 1, 2, 2, 2, 0, 0, 2, 2, 2, 0, 0, 0, 1, 0, 1, 2, 2]
+        self.combining_tensor = self.set_tensor_values(tensor=self.combining_tensor, matrix=matrix_numbers,
+                                                       column=column_numbers,
+                                                       row=row_numbers, values=values)
+        self.combining_tensor = torch.from_numpy(self.combining_tensor)
+
+        self.combining_bias = torch.from_numpy(np.full(shape=(3,), fill_value=0.0, dtype='float32'))
+
+    @staticmethod
+    def set_tensor_values(tensor, matrix, column, row, values):
+        """
+        Helper method to initialize a third-order tensor with specific values
+        """
+        for i in range(len(matrix)):
+            tensor[matrix[i]][column[i]][row[i]] = values[i]
+        return tensor
+
+    def test_transformation(self):
+        """
```

Comment:
it would be nice to have an explanation of the tensor contraction as present below for the weighting function